warehouse/query-core/src/main/java/datawave/query/iterator/QueryLogIterator.java

```diff
@@ -0,0 +1,191 @@
+import org.apache.log4j.Logger;
+
+/**
+ * An iterator used to log the start and end of each method run by the {@link SortedKeyValueIterator} above it in the iterator stack. Logs the QueryID
+ * associated with each method. Logs are written on the TServer where the iterator is running
+ */
+public class QueryLogIterator implements SortedKeyValueIterator<Key,Value>, OptionDescriber {
+
+    private static final Logger log = Logger.getLogger(QueryLogIterator.class);
+    private static final String CLASS_NAME = QueryLogIterator.class.getSimpleName();
+
+    private String queryID;
+    private SortedKeyValueIterator<Key,Value> source;
+    private IteratorEnvironment env;
+
+    /**
+     * Default constructor
+     */
+    public QueryLogIterator() {
+        // no-arg constructor
+    }
+
+    /**
+     * Class copy constructor
+     */
+    public QueryLogIterator(QueryLogIterator other, IteratorEnvironment env) {
+        this.source = other.source.deepCopy(env);
+        this.env = other.env;
+        this.queryID = other.queryID;
+    }
+
+    /**
+     * Wraps the init() method of the iterator above it, logging the start/end of the method along with its query id.
+     */
+    @Override
+    public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options, IteratorEnvironment env) throws IOException {
+
+        try {
+            logStartOf("init()");
+            this.queryID = options.get(QUERY_ID);
+            this.source = source;
+            this.env = env;
+        } finally {
+            logEndOf("init()");
+        }
+    }
+
+    /**
+     * Logs the query id and {@link SortedKeyValueIterator} method name before the method is run.
+     */
+    private void logStartOf(String methodName) {
+        if (log.isDebugEnabled()) {
+            log.debug(CLASS_NAME + " " + methodName + " Started QueryID: " + this.queryID);
+        }
+    }
+
+    /**
+     * Logs the query id and {@link SortedKeyValueIterator} method name after the method is run.
+     */
+    private void logEndOf(String methodName) {
+        if (log.isDebugEnabled()) {
+            log.debug(CLASS_NAME + " " + methodName + " Ended QueryID: " + this.queryID);
+        }
+    }
+
+    /**
+     * Wraps the hasTop() method of the iterator above it, logging the start and end of the method along with its query id.
+     */
+    @Override
+    public boolean hasTop() {
+        boolean result;
+
+        try {
+            logStartOf("hasTop()");
+            result = source.hasTop();
+        } finally {
+            logEndOf("hasTop()");
+        }
+        return result;
+    }
+
+    /**
+     * Wraps the next() method of the iterator above it, logging the start and end of the method along with its query id.
+     */
+    @Override
+    public void next() throws IOException {
+        try {
+            logStartOf("next()");
+            source.next();
+        } finally {
+            logEndOf("next()");
+        }
+    }
```

Comment:
Would like to see changing the thread name and adding queryId to a MappedDiagnosticContext